src/runtime/io.cpp

```diff
@@ -89,7 +89,14 @@ static obj_res mk_file_not_found_error(b_obj_arg fname) {
 static lean_external_class * g_io_handle_external_class = nullptr;
 
 static void io_handle_finalizer(void * h) {
-    lean_always_assert(fclose(static_cast<FILE *>(h)) == 0);
+    int errnum = fclose(static_cast<FILE *>(h));
+    if (errnum != 0) {
+        object * str = lean_io_error_to_string(decode_io_error(errnum, nullptr));
+        sstream ss;
+        ss << "could not close finalized `IO.FS.Handle`: " << string_cstr(str);
+        lean_internal_panic(ss.str().c_str());
+        dec_ref(str);
+    }
 }
```

Comment:
I was surprised to see there is no `IO.FS.Handle.close` operation.  That seems like it'd be sensible to add, and then this code would need to handle that case (`fclose` returning non-zero and then checking if `errno` is `ENOTOPEN`).

I could add that as a separate commit though.

Reply:
Yes, sounds good to me as a follow-up PR